src/git/http.ts

```diff
@@ -0,0 +1,408 @@
+      const [type, objectId, parsedCapabilities, rest] = parsedData;
+      workingBuffer = rest;
+      if (firstLine) {
+        capabilities.push(...parsedCapabilities);
+        firstLine = false;
+      }
+      switch (type) {
+        case 'want':
+          wants.push(objectId);
+          break;
+        case 'have':
+          haves.push(objectId);
+          break;
+        case 'SEPARATOR':
+          break;
+        case 'done':
+          return [wants, haves, capabilities];
+        default:
+          utils.never(
+            `Type should be either 'want' or 'have', found '${type}'`,
+          );
+      }
+    }
+  }
+  return [wants, haves, capabilities];
+}
+
+/**
+ * Parses the client's requests and generates a response the contains the git packFile data
+ *
+ * It will respond with the `PKT-LINE(NAK_BUFFER)` and then the `packFile` data chunked into lines for the stream.
+ *
+ */
+async function* generatePackRequest({
+  fs,
+  dir,
+  gitDir,
+  body,
+}: {
+  fs: EncryptedFS;
+  dir: string;
+  gitDir: string;
+  body: Array<Buffer>;
+}): AsyncGenerator<Buffer, void, void> {
+  const [wants, haves, _capabilities] = await parsePackRequest(body);
+  const objectIds = await gitUtils.listObjects({
+    fs,
+    dir,
+    gitDir: gitDir,
+    wants,
+    haves,
+  });
+  // Reply that we have no common history and that we need to send everything
+  yield packetLineBuffer(gitUtils.NAK_BUFFER);
+  // Send everything over in pack format
+  yield* generatePackData({
+    fs,
+    dir,
+    gitDir,
+    objectIds,
+  });
+  // Send dummy progress data
+  yield packetLineBuffer(
+    gitUtils.DUMMY_PROGRESS_BUFFER,
+    gitUtils.CHANNEL_PROGRESS,
+  );
+  // Send flush
+  yield gitUtils.FLUSH_PACKET_BUFFER;
+}
+
+/**
+ * Generates the packFile data to be included within the response.
+ * Iso-git provides the packFile for us, we just need to cut it into `chunkSize` bytes per line and multiplex on chanel 1.
+ *
+ */
+async function* generatePackData({
+  fs,
+  dir,
+  gitDir,
+  objectIds,
+  chunkSize = gitUtils.PACK_CHUNK_SIZE,
+}: {
+  fs: EncryptedFS;
+  dir: string;
+  gitDir: string;
+  objectIds: Array<ObjectId>;
+  chunkSize?: number;
+}): AsyncGenerator<Buffer, void, void> {
+  const packFile = await git.packObjects({
+    fs,
+    dir,
+    gitdir: gitDir,
+    oids: objectIds,
+  });
+  if (packFile.packfile == null) utils.never('failed to create packFile data');
+  let packFileBuffer = Buffer.from(packFile.packfile.buffer);
+
+  // Streaming the packFile as chunks of the length specified by the `chunkSize`.
+  // Each line is formatted as a `PKT-LINE`
+  do {
+    const subBuffer = packFileBuffer.subarray(0, chunkSize);
+    packFileBuffer = packFileBuffer.subarray(chunkSize);
+    yield packetLineBuffer(subBuffer, gitUtils.CHANNEL_DATA);
+  } while (packFileBuffer.byteLength > chunkSize);
+}
+
+export {
+  advertiseRefGenerator,
+  packetLineBuffer,
+  parsePackRequest,
+  generatePackRequest,
+  generatePackData,
```

Comment:
I would like docblocks here to make it clear exactly the dependency tree, or call graph of the http protocol itself.

As in what is considered the main function in relation to the cloning and pulling operations? And how does it decompose subsequently. Docblocks should use ASCII diagrams to make this clear.

Reply:
Updated the relevant docblocks.